File: SkillForge-AI-Driven-Adaptive-Learning-Exam-Generator-main/SkillForge-AI-Driven-ExamGenerator-Project-main/BackEnd/SkillForge_1/src/main/java/com/example/SkillForge_1/service/AnalyticsService.java
package com.example.SkillForge_1.service;

import com.example.SkillForge_1.dto.CoursePerformanceDTO;
import com.example.SkillForge_1.dto.StudentQuizScoreDTO;
import com.example.SkillForge_1.repository.AnalyticsRepository;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Map;

@Service
public class AnalyticsService {

    private final AnalyticsRepository analyticsRepository;

    public AnalyticsService(AnalyticsRepository analyticsRepository) {
        this.analyticsRepository = analyticsRepository;
    }

    // 1️⃣ Student Quiz Scores
    public List<StudentQuizScoreDTO> getStudentQuizScores() {
        List<Object[]> rawData = analyticsRepository.getStudentQuizScores();
        List<StudentQuizScoreDTO> result = new java.util.ArrayList<>();

        for (Object[] row : rawData) {
            Long studentId = (Long) row[0];
            String studentName = (String) row[1];
            Long quizId = (Long) row[2];
            String quizTitle = (String) row[3];
            String topicName = (String) row[4];
            String difficulty = (String) row[5];
            
            // Handle potentially null numbers from DB
            Double score = row[6] != null ? ((Number) row[6]).doubleValue() : 0.0;
            Integer totalMarks = row[7] != null ? ((Number) row[7]).intValue() : 0;
            java.time.LocalDateTime submittedAt = (java.time.LocalDateTime) row[8];
            String status = (String) row[9];

            Double percentage = (totalMarks > 0 && score > 0) ? (score * 100.0 / totalMarks) : 0.0;

            result.add(new StudentQuizScoreDTO(
                studentId, studentName, quizId, quizTitle, topicName, difficulty,
                score, totalMarks, percentage, status, submittedAt
            ));
        }
        return result;
    }


    // 2️⃣ Course Performance
    public List<CoursePerformanceDTO> getCoursePerformance() {
        // Fetch raw data: [CourseName, StudentId, StudentName, TotalScore, TotalMaxMarks]
        List<Object[]> rawData = analyticsRepository.getRawCoursePerformanceData();
        
        Map<String, CoursePerformanceDTO> courseMap = new java.util.HashMap<>();

        for (Object[] row : rawData) {
            String courseName = (String) row[0];
            Long studentId = (Long) row[1];
            String studentName = (String) row[2];
            Long totalScore = ((Number) row[3]).longValue();
            Long totalMarks = ((Number) row[4]).longValue();

            double percentage = (totalMarks > 0) ? (totalScore * 100.0 / totalMarks) : 0.0;

            // Get or create course DTO
            CoursePerformanceDTO courseDTO = courseMap.computeIfAbsent(courseName, k -> new CoursePerformanceDTO(k, 0.0));
            
            // Add student performance
            courseDTO.getStudents().add(new CoursePerformanceDTO.StudentPerformanceDTO(studentId, studentName, percentage));
        }

        // Calculate average for each course and finalize list
        List<CoursePerformanceDTO> result = new java.util.ArrayList<>(courseMap.values());
        for (CoursePerformanceDTO dto : result) {
            if (!dto.getStudents().isEmpty()) {
                double avg = dto.getStudents().stream()
                        .mapToDouble(CoursePerformanceDTO.StudentPerformanceDTO::getPercentage)
                        .average()
                        .orElse(0.0);
                dto.setAverageScore(avg);
            }
        }
        
        return result;
    }


    // 3️⃣ Topic Performance
    public List<Map<String, Object>> getTopicPerformance() {
        return analyticsRepository.getTopicPerformance();
    }

    // 4️⃣ Adaptive Skill Progression
    public List<Map<String, Object>> getSkillProgression() {
        return analyticsRepository.getSkillProgression();
    }
    
    // 5️⃣ Daily Student Involvement
    public List<Map<String, Object>> getDailyStudentInvolvement() {
        return analyticsRepository.getDailyStudentInvolvement();
    }
    
    // 6️⃣ Student Performance by Course
    public List<Map<String, Object>> getStudentPerformanceByCourse(String courseName) {
        return analyticsRepository.getStudentPerformanceByCourse(courseName);
    }
    
    // 7️⃣ Overall Course Performance
    public List<Map<String, Object>> getOverallCoursePerformance() {
        return analyticsRepository.getOverallCoursePerformance();
    }
    
    // 5️⃣ Student Detailed Analytics
    public Map<String, Object> getStudentDetailedAnalytics(Long studentId) {
        Map<String, Object> result = new java.util.HashMap<>();
        
        // Student info
        Map<String, Object> studentInfo = analyticsRepository.getStudentInfo(studentId);
        if (studentInfo == null) {
            studentInfo = new java.util.HashMap<>();
            studentInfo.put("id", studentId);
            studentInfo.put("name", "Student " + studentId);
            studentInfo.put("email", "student" + studentId + "@example.com");
        }
        result.put("studentInfo", studentInfo);
        
        // Course performance for this student
        List<Map<String, Object>> coursePerf = analyticsRepository.getStudentCoursePerformance(studentId);
        result.put("coursePerformance", coursePerf);
        
        // Topic performance for this student
        List<Map<String, Object>> topicPerf = analyticsRepository.getStudentTopicPerformance(studentId);
        result.put("topicPerformance", topicPerf);
        
        // Activity metrics
        Map<String, Object> metricsRaw = analyticsRepository.getStudentActivityMetrics(studentId);
        Map<String, Object> metrics = new java.util.HashMap<>();
        
        Long totalAssigned = metricsRaw.get("totalAssigned") != null 
            ? ((Number) metricsRaw.get("totalAssigned")).longValue() : 0L;
        Long totalAttempted = metricsRaw.get("totalAttempted") != null 
            ? ((Number) metricsRaw.get("totalAttempted")).longValue() : 0L;
        
        metrics.put("totalAssigned", totalAssigned);
        metrics.put("totalAttempted", totalAttempted);
        metrics.put("completionRate", totalAssigned > 0 
            ? Math.round((totalAttempted * 100.0 / totalAssigned)) : 0);
        result.put("activityMetrics", metrics);
        
        // Recent attempts
        List<Map<String, Object>> recentAttempts = analyticsRepository.getStudentRecentAttempts(studentId);
        result.put("recentAttempts", recentAttempts);
        
        return result;
    }
    
    // 8️⃣ Get All Students
    public List<Map<String, Object>> getAllStudents() {
        return analyticsRepository.getAllStudents();
    }
    
    // 10️⃣ Quiz Analytics
    public Map<String, Object> getQuizAnalytics(Long quizId) {
        Map<String, Object> basicAnalytics = analyticsRepository.getQuizAnalytics(quizId);
        
        // Get detailed attempts and assignments
        List<Map<String, Object>> attempts = getQuizAttempts(quizId);
        List<Map<String, Object>> assignedStudents = getQuizAssignments(quizId);
        
        // Merge basic analytics with detailed data
        Map<String, Object> result = new java.util.HashMap<>(basicAnalytics);
        result.put("attempts", attempts);
        result.put("assignedStudents", assignedStudents);
        
        return result;
    }
    
    // Helper method to get quiz attempts
    private List<Map<String, Object>> getQuizAttempts(Long quizId) {
        return analyticsRepository.getQuizAttempts(quizId);
    }
    
    // Helper method to get quiz assignments
    private List<Map<String, Object>> getQuizAssignments(Long quizId) {
        return analyticsRepository.getQuizAssignments(quizId);
    }
    
    // 9️⃣ Get Student ID by Username
    public Long getStudentIdByUsername(String username) {
        return analyticsRepository.getStudentIdByUsername(username);
    }
}

File: SkillForge-AI-Driven-Adaptive-Learning-Exam-Generator-main/SkillForge-AI-Driven-ExamGenerator-Project-main/FrontEnd-Antigravity/Anti-start/src/services/analyticsService.js
import axios from 'axios';

const API_BASE_URL = 'http://localhost:8081/api';

// Get auth token from localStorage
const getAuthHeader = () => {
    const token = localStorage.getItem('skillforge_token');
    return token ? { Authorization: `Bearer ${token}` } : {};
};

/**
 * Analytics Service
 * Provides comprehensive performance analytics for instructors
 * All methods include detailed console logging for debugging
 */
const analyticsService = {
    /**
     * Get course-wise student performance distribution
     * Shows how students performed across different courses
     */
    getCoursePerformance: async () => {
        console.log('📤 [Analytics API] Fetching course-wise student performance');

        try {
            const response = await axios.get(`${API_BASE_URL}/analytics/course-performance`, {
                headers: getAuthHeader()
            });

            console.log('📥 [Analytics API] Received course performance data:', response.data);
            return response.data;
        } catch (error) {
            console.error('❌ [Analytics API] Error fetching course performance:', error);
            return [];
        }
    },

    /**
     * Get topic-wise performance across all courses
     * Shows average scores for each topic
     */
    getTopicPerformance: async () => {
        console.log('📤 [Analytics API] Fetching topic-wise performance');

        try {
            const response = await axios.get(`${API_BASE_URL}/analytics/topic-performance`, {
                headers: getAuthHeader()
            });

            console.log('📥 [Analytics API] Received topic performance data:', response.data);
            return response.data;
        } catch (error) {
            console.error('❌ [Analytics API] Error fetching topic performance:', error);
            return [];
        }
    },

    /**
     * Get adaptive skill progression over time
     * Shows how student performance changes as difficulty increases
     */
    getSkillProgression: async () => {
        console.log('📤 [Analytics API] Fetching adaptive skill progression');

        try {
            const response = await axios.get(`${API_BASE_URL}/analytics/skill-progression`, {
                headers: getAuthHeader()
            });

            console.log('📥 [Analytics API] Received skill progression data:', response.data);
            return response.data;
        } catch (error) {
            console.error('❌ [Analytics API] Error fetching skill progression:', error);
            return [];
        }
    },
    
    /**
     * Get daily student involvement
     */
    getDailyStudentInvolvement: async () => {
        console.log('📤 [Analytics API] Fetching daily student involvement');

        try {
            const response = await axios.get(`${API_BASE_URL}/analytics/daily-involvement`, {
                headers: getAuthHeader()
            });

            console.log('📥 [Analytics API] Received daily involvement data:', response.data);
            return response.data;
        } catch (error) {
            console.error('❌ [Analytics API] Error fetching daily involvement:', error);
            return [];
        }
    },
    
    /**
     * Get student performance for a specific course
     */
    getStudentPerformanceByCourse: async (courseName) => {
        console.log('📤 [Analytics API] Fetching student performance for course:', courseName);

        try {
            const response = await axios.get(`${API_BASE_URL}/analytics/course/${encodeURIComponent(courseName)}/students`, {
                headers: getAuthHeader()
            });

            console.log('📥 [Analytics API] Received course student data:', response.data);
            return response.data;
        } catch (error) {
            console.error('❌ [Analytics API] Error fetching course student data:', error);
            return [];
        }
    },
    
    /**
     * Get overall course performance
     */
    getOverallCoursePerformance: async () => {
        console.log('📤 [Analytics API] Fetching overall course performance');

        try {
            const response = await axios.get(`${API_BASE_URL}/analytics/overall-course-performance`, {
                headers: getAuthHeader()
            });

            console.log('📥 [Analytics API] Received overall course performance:', response.data);
            return response.data;
        } catch (error) {
            console.error('❌ [Analytics API] Error fetching overall course performance:', error);
            return [];
        }
    },

    /**
     * Get student-quiz scores table data
     * Supports filtering by course, topic, and difficulty
     */
    getStudentQuizScores: async (filters = {}) => {
        console.log('📤 [Analytics API] Fetching student-quiz scores', filters);

        try {
            const params = new URLSearchParams();
            if (filters.studentName) params.append('studentName', filters.studentName);
            if (filters.topicId) params.append('topic', filters.topicId);
            if (filters.difficulty) params.append('difficulty', filters.difficulty);

            const response = await axios.get(`${API_BASE_URL}/analytics/student-quiz-scores?${params.toString()}`, {
                headers: getAuthHeader()
            });

            console.log('📥 Received student-quiz scores:', response.data);
            return response.data;
        } catch (error) {
            console.error('❌ [Analytics API] Error fetching student-quiz scores:', error);
            return [];
        }
    },

    /**
     * Get detailed analytics for a specific student
     * Includes course performance, topic performance, activity metrics, and recent attempts
     */
    getStudentDetailedAnalytics: async (studentId) => {
        console.log('📤 [Analytics API] Fetching detailed analytics for student:', studentId);

        try {
            const response = await axios.get(`${API_BASE_URL}/analytics/student/${studentId}/detailed`, {
                headers: getAuthHeader()
            });

            console.log('📥 [Analytics API] Received detailed student analytics:', response.data);
            return response.data;
        } catch (error) {
            console.error('❌ [Analytics API] Error fetching detailed student analytics:', error);
            return {
                studentInfo: { id: studentId, name: 'Unknown', email: '' },
                coursePerformance: [],
                topicPerformance: [],
                activityMetrics: { totalAssigned: 0, totalAttempted: 0, completionRate: 0 },
                recentAttempts: []
            };
        }
    },

    /**
     * Get all students with basic information
     * Returns list of students with id, name, and email
     */
    getAllStudents: async () => {
        console.log('📤 [Analytics API] Fetching all students');

        try {
            const response = await axios.get(`${API_BASE_URL}/analytics/students`, {
                headers: getAuthHeader()
            });

            console.log('📥 [Analytics API] Received all students:', response.data);
            return response.data;
        } catch (error) {
            console.error('❌ [Analytics API] Error fetching all students:', error);
            return [];
        }
    },

    /**
     * Get current student's own performance analytics
     * Uses authentication to fetch only their own data
     */
    getMyPerformance: async () => {
        console.log('📤 [Analytics API] Fetching my performance');

        try {
            const response = await axios.get(`${API_BASE_URL}/analytics/my-performance`, {
                headers: getAuthHeader()
            });

            console.log('📥 [Analytics API] Received my performance:', response.data);
            return response.data;
        } catch (error) {
            console.error('❌ [Analytics API] Error fetching my performance:', error);
            return {
                studentInfo: { id: 0, name: 'Unknown', email: '' },
                coursePerformance: [],
                topicPerformance: [],
                activityMetrics: { totalAssigned: 0, totalAttempted: 0, completionRate: 0 },
                recentAttempts: []
            };
        }
    }
};

export default analyticsService;
